File: titanic.py
#%%
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

teste = pd.read_csv(r'inputs\test.csv', sep=',')
treinamento = pd.read_csv(r'inputs\train.csv', sep=',')
treinamento.columns = treinamento.columns.str.upper()
teste.columns = teste.columns.str.upper()
#%%
print(treinamento.nunique())
print(treinamento.isnull().sum().sort_values(ascending=False))
print(treinamento.dtypes)
#%%
#juntando df de treino e teste
df_merged = pd.concat([treinamento, teste]).reset_index(drop=True)
#preenchendo idades faltantes com a mediana
mediana_idade = df_merged.AGE.median()
df_merged['AGE'] = df_merged.AGE.fillna(mediana_idade)
#preenchendo tarifas faltantes com a mediana
mediana_tarifa = df_merged.FARE.median()
df_merged['FARE'] = df_merged.FARE.fillna(mediana_tarifa)
#transformando sexo em 0 - masculino e 1 - feminino
df_merged['SEX'] = df_merged.SEX.map({'male':0,
                                      'female':1})
#separando teste e treinamento
treinamento = df_merged.iloc[:treinamento.shape[0]]
teste = df_merged.iloc[treinamento.shape[0]:]
#%%
#MODELO COM REGRESSÃO LOGISTICA
# criar um modelo de Regressão Logística
#separando variaveis explicativas do modelo e alvo
var_treinamento = treinamento[['PCLASS', 'SEX', 'FARE','AGE','SIBSP','PARCH']]
var_teste = teste[['PCLASS', 'SEX', 'FARE','AGE','SIBSP','PARCH']]
alvo = treinamento.SURVIVED
#fit modelo
modelo_log = LogisticRegression(solver='liblinear')
modelo_log.fit(var_treinamento, alvo)

# verificar a acurácia do modelo
acuracia_log = round(modelo_log.score(var_treinamento, alvo) * 100, 2)
print("Acurácia do modelo de Regressão Logística: {}".format(acuracia_log))
#rodando conjunto teste
modelo_teste = modelo_log.predict(var_teste)
resultado = pd.DataFrame({'PassengerID':teste.PASSENGERID,
                         'Survived':modelo_teste})
resultado.to_csv(r'outputs\result_lr.csv', index = False, sep=';')
#%%
# criar um modelo de árvore de decisão
modelo_arvore = DecisionTreeClassifier(max_depth=3)
modelo_arvore = modelo_arvore.fit(var_treinamento, alvo)

# verificar a acurácia do modelo
acuracia_arvore = round(modelo_arvore.score(var_treinamento, alvo) * 100, 2)
print("Acurácia do modelo de Árvore de Decisão: {}".format(acuracia_arvore))
arvore_teste = modelo_arvore.predict(var_teste)
resultado_arvore = pd.DataFrame({'PassengerID':teste.PASSENGERID,
                         'Survived':arvore_teste})
resultado_arvore.to_csv(r'outputs\result_tree.csv', index = False, sep=';')
#%%
resultado_train = modelo_log.predict(var_treinamento)
matriz_confusao = confusion_matrix(treinamento['SURVIVED'], resultado_train)
matriz_confusao
# %%
resultado_train = modelo_arvore.predict(var_treinamento)
matriz_confusao = confusion_matrix(treinamento['SURVIVED'], resultado_train)
matriz_confusao
#%%
# Exibir a matriz de confusão com melhores resultados usando Seaborn
sns.heatmap(matriz_confusao, annot=True, fmt="d", cmap="Blues")
plt.xlabel('Predito')
plt.ylabel('Real')
plt.title('Matriz de Confusão')
plt.show()

File: titanic.r
library(tidyverse)
library(dplyr)
library(gmodels)
library(pheatmap)
library(rpart)


#código previamente feito em python, refazendo em R pontos cruciais
train <- read.csv('inputs\\train.csv')
test <- read.csv('inputs\\test.csv')

df_merged <- bind_rows(train, test)
#substituindo nulos na idade

df_merged$Age[is.na(df_merged$Age)] <- median(df_merged$Age, na.rm = TRUE)
#substituindo nulos na tarifa
df_merged$Fare[is.na(df_merged$Fare)] <- median(df_merged$Fare, na.rm = TRUE)
#alterando sexo para 0 - masculino e 1 - feminino
mapeamento <- c("male" = 0, "female" = 1)
df_merged$Sex <- mapeamento[df_merged$Sex]
#separando novamente treino e teste
train_trat <- df_merged[1:nrow(train),]
test_trat <- df_merged[(nrow(train)+1):nrow(df_merged),]
#regressão logistica
modelo_log <- glm(formula = Survived ~ Pclass + Sex + Fare + Age + SibSp + Parch, data = train_trat, family = binomial)
resultados_log <- predict(modelo_log, test_trat, type = "response")
#transformando em binario
resultados_log <- ifelse(resultados_log > 0.5, 1, 0)
final_log <- test %>% select(PassengerId) %>% mutate(Survived = resultados_log)
#arvore de decisao
modelo_arvore <- rpart(Survived ~ Pclass + Sex + Fare + Age + SibSp + Parch, data = train_trat, method = "class")
# Fazer previsões no conjunto de teste
resultados_arvore <- predict(modelo_arvore, newdata = test_trat, type = "class")
final_arvore <- test %>% select(PassengerId) %>% mutate(Survived = resultados_arvore)
# Avaliar a precisão do modelo
#matriz confusão com modelo de treinamento
treino_log <- predict(modelo_log, train_trat, type = "response")
treino_log <- ifelse(treino_log > 0.5, 1, 0)
matriz_confusao_lg <- table(train_trat$Survived, treino_log)
#matriz confusão com modelo de treinamento
treino_arvore <- predict(modelo_arvore, train_trat, type = "class")
matriz_confusao_av <- table(train_trat$Survived, treino_arvore)
#salvando arquivos
write.csv2(final_log, "outputs/r_result_lr.csv", row.names = FALSE)
write.csv2(final_arvore, "outputs/r_result_tree.csv", row.names = FALSE)
